File: app/src/main/java/com/example/homestay/AdminDashboardActivity.java
package com.example.homestay;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.PopupMenu;
import com.example.homestay.data.entity.*;
import com.example.homestay.data.repository.HomestayRepository;
import com.example.homestay.utils.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.Executors;

public class AdminDashboardActivity extends AppCompatActivity {
  private HomestayRepository repository;

  @Override
  protected void onCreate(Bundle s) {
    super.onCreate(s);
    setContentView(R.layout.activity_admin_dashboard);
    SystemBarUtils.keepContentBelowStatusBar(this);
    if (!isAdminLoggedIn()) {
      finish();
      return;
    }
    repository = ((HomestayApplication) getApplication()).getRepository();
    setupViews();
    AdminNavigationUtils.setup(this, R.id.admin_nav_dashboard);
    ensureAdminCustomerSession();
  }

  @Override
  protected void onResume() {
    super.onResume();
    if (repository != null) loadDashboard();
  }

  private void loadDashboard() {
    Executors.newSingleThreadExecutor()
        .execute(
            () -> {
              List<Room> rooms = repository.getAllRoomsNow();
              List<User> users = repository.getAllUsersNow();
              List<Booking> bookings = repository.getAllBookingsNow();
              runOnUiThread(() -> render(rooms, users, bookings));
            });
  }

  private void render(List<Room> rooms, List<User> users, List<Booking> bookings) {
    int pending = 0, confirmed = 0, completed = 0, cancelled = 0, active = 0;
    double expected = 0, actual = 0;
    long now = System.currentTimeMillis();
    Set<Long> occupied = new HashSet<>();
    for (Room r : rooms) if (r.isAvailable()) active++;
    for (Booking b : bookings) {
      switch (b.getStatus()) {
        case "pending":
          pending++;
          break;
        case "confirmed":
          confirmed++;
          break;
        case "completed":
          completed++;
          break;
        case "cancelled":
          cancelled++;
          break;
      }
      if ("confirmed".equals(b.getStatus()) || "completed".equals(b.getStatus()))
        expected += b.getTotalPrice();
      if ("completed".equals(b.getStatus())) actual += b.getTotalPrice();
      if ("confirmed".equals(b.getStatus())
          && b.getCheckInDate() <= now
          && b.getCheckOutDate() >= now) occupied.add(b.getRoomId());
    }
    text(R.id.tv_stat_rooms, rooms.size() + "\nPhòng");
    text(R.id.tv_stat_users, users.size() + "\nNgười dùng");
    text(R.id.tv_stat_pending, pending + "\nChờ duyệt");
    text(R.id.tv_stat_revenue, "Doanh thu dự kiến: " + money(expected) + " đ");
    text(R.id.tv_stat_revenue_actual, "Doanh thu thực tế: " + money(actual) + " đ");
    text(R.id.tv_booking_breakdown, bookings.size() + "\nTổng đặt chỗ");
    text(R.id.tv_booking_completed, completed + "\nHoàn thành");
    text(R.id.tv_booking_cancelled, cancelled + "\nĐã hủy");
    Calendar cal = Calendar.getInstance();
    cal.set(Calendar.DAY_OF_MONTH, 1);
    cal.set(Calendar.HOUR_OF_DAY, 0);
    int fresh = 0;
    for (User u : users) if (u.getCreatedAt() >= cal.getTimeInMillis()) fresh++;
    int occupancy = active == 0 ? 0 : occupied.size() * 100 / active;
    text(
        R.id.tv_operational_summary,
        "Phòng đang mở     "
            + active
            + "/"
            + rooms.size()
            + "        Đang lưu trú     "
            + occupied.size()
            + "\nNgười dùng mới     "
            + fresh
            + "        Lấp đầy hiện tại     "
            + occupancy
            + "%");
    LinearLayout container = findViewById(R.id.recent_bookings_container);
    container.removeAllViews();
    bookings.sort((a, b) -> Long.compare(b.getCreatedAt(), a.getCreatedAt()));
    SimpleDateFormat date = new SimpleDateFormat("dd/MM HH:mm", new Locale("vi", "VN")),
        codeDate = new SimpleDateFormat("yyMMdd", Locale.getDefault());
    for (int i = 0; i < Math.min(3, bookings.size()); i++) {
      Booking b = bookings.get(i);
      View card = getLayoutInflater().inflate(R.layout.item_admin_recent_booking, container, false);
      ((TextView) card.findViewById(R.id.tv_recent_code))
          .setText(
              "#RG"
                  + codeDate.format(new Date(b.getCreatedAt()))
                  + String.format(Locale.ROOT, "%03d", b.getId()));
      TextView status = card.findViewById(R.id.tv_recent_status);
      status.setText(statusLabel(b.getStatus()));
      status.setTextColor(statusColor(b.getStatus()));
      ((TextView) card.findViewById(R.id.tv_recent_meta))
          .setText(date.format(new Date(b.getCreatedAt())));
      ((TextView) card.findViewById(R.id.tv_recent_price)).setText(money(b.getTotalPrice()) + " đ");
      container.addView(card);
    }
  }

  private void setupViews() {
    TextView name = findViewById(R.id.tv_admin_name);
    name.setText(
        getSharedPreferences("AdminSession", MODE_PRIVATE)
            .getString("admin_fullname", "Administrator"));
    int[] hidden = {
      R.id.card_manage_rooms,
      R.id.card_view_users,
      R.id.card_manage_bookings,
      R.id.btn_logout,
      R.id.btn_switch_to_guest,
      R.id.tv_management_title
    };
    for (int id : hidden) findViewById(id).setVisibility(View.GONE);
    findViewById(R.id.btn_admin_account)
        .setOnClickListener(
            anchor -> {
              PopupMenu menu = new PopupMenu(this, anchor);
              menu.getMenu()
                  .add("Xem giao diện khách")
                  .setOnMenuItemClickListener(
                      x -> {
                        switchToGuest();
                        return true;
                      });
              menu.getMenu()
                  .add("Đăng xuất")
                  .setOnMenuItemClickListener(
                      x -> {
                        logout();
                        return true;
                      });
              menu.show();
            });
  }

  private void ensureAdminCustomerSession() {
    Executors.newSingleThreadExecutor()
        .execute(
            () -> {
              SessionManager session = new SessionManager(this);
              if (session.isLoggedIn()) return;
              User user = repository.getUserByEmail(AdminAuth.EMAIL);
              if (user == null) {
                long id =
                    repository.insertUser(
                        new User(
                            0,
                            AdminAuth.EMAIL,
                            "admin-local",
                            "AdminLocalSession@1",
                            "Administrator",
                            System.currentTimeMillis()));
                user = repository.getUserById(id);
              }
              if (user != null)
                session.saveSession(
                    user.getId(), "local-admin", user.getEmail(), user.getFullName());
            });
  }

  private void switchToGuest() {
    ensureAdminCustomerSession();
    Intent i = new Intent(this, MainActivity.class);
    i.putExtra("admin_guest_preview", true);
    i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
    startActivity(i);
    finish();
  }

  private void logout() {
    getSharedPreferences("AdminSession", MODE_PRIVATE).edit().clear().apply();
    new SessionManager(this).clearSession();
    Intent i = new Intent(this, LoginActivity.class);
    i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
    startActivity(i);
    finish();
  }

  private boolean isAdminLoggedIn() {
    return getSharedPreferences("AdminSession", MODE_PRIVATE)
        .getBoolean("is_admin_logged_in", false);
  }

  private void text(int id, String value) {
    ((TextView) findViewById(id)).setText(value);
  }

  private static String money(double v) {
    return NumberFormat.getNumberInstance(new Locale("vi", "VN")).format((long) v);
  }

  private static String statusLabel(String s) {
    switch (s) {
      case "pending":
        return "Chờ duyệt";
      case "confirmed":
        return "Đã xác nhận";
      case "completed":
        return "Hoàn thành";
      case "cancelled":
        return "Đã hủy";
      default:
        return s;
    }
  }

  private static int statusColor(String s) {
    switch (s) {
      case "pending":
        return 0xFFA55A00;
      case "confirmed":
        return 0xFF167A50;
      case "completed":
        return 0xFF0B4AA2;
      case "cancelled":
        return 0xFFD14343;
      default:
        return 0xFF737B8C;
    }
  }
}

File: app/src/main/java/com/example/homestay/AdminDashboardActivity.kt
package com.example.homestay

import android.content.Intent
import android.os.Bundle
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.lifecycle.Lifecycle
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.launch
import java.text.NumberFormat
import java.util.Locale
import java.text.SimpleDateFormat
import java.util.Date
import com.google.android.material.button.MaterialButton
import com.google.android.material.card.MaterialCardView
import com.example.homestay.utils.SessionManager
import com.example.homestay.utils.AdminAuth
import com.example.homestay.data.entity.User

class AdminDashboardActivity : AppCompatActivity() {
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_admin_dashboard)
        
        // Check if admin logged in
        if (!isAdminLoggedIn()) {
            finish()
            return
        }
        
        setupViews()
        observeDashboard()
        lifecycleScope.launch { ensureAdminCustomerSession() }
    }

    private fun observeDashboard() {
        val repository = (application as HomestayApplication).repository
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                combine(
                    repository.getAllRooms(),
                    repository.getAllUsers(),
                    repository.getAllBookings()
                ) { rooms, users, bookings -> Triple(rooms, users, bookings) }
                    .collect { (rooms, users, bookings) ->
                        findViewById<TextView>(R.id.tv_stat_rooms).text = "${rooms.size}\nPhòng"
                        findViewById<TextView>(R.id.tv_stat_users).text = "${users.size}\nNgười dùng"
                        findViewById<TextView>(R.id.tv_stat_pending).text = "${bookings.count { it.status == "pending" }}\nChờ duyệt"
                        val revenue = bookings.filter { it.status in setOf("confirmed", "completed") }.sumOf { it.totalPrice }
                        val amount = NumberFormat.getNumberInstance(Locale("vi", "VN")).format(revenue.toLong())
                        findViewById<TextView>(R.id.tv_stat_revenue).text = "Doanh thu dự kiến: $amount đ"
                        val actualRevenue = bookings.filter { it.status == "completed" }.sumOf { it.totalPrice }
                        val actualAmount = NumberFormat.getNumberInstance(Locale("vi", "VN")).format(actualRevenue.toLong())
                        findViewById<TextView>(R.id.tv_stat_revenue_actual).text = "Doanh thu thực tế: $actualAmount đ"
                        val confirmed = bookings.count { it.status == "confirmed" }
                        val completed = bookings.count { it.status == "completed" }
                        val cancelled = bookings.count { it.status == "cancelled" }
                        findViewById<TextView>(R.id.tv_booking_breakdown).text =
                            "Tổng booking: ${bookings.size}\nĐã xác nhận: $confirmed  •  Hoàn thành: $completed  •  Đã hủy: $cancelled"
                        val activeRooms = rooms.count { it.isAvailable }
                        val occupiedRoomIds = bookings.filter { it.status == "confirmed" && it.checkInDate <= System.currentTimeMillis() && it.checkOutDate >= System.currentTimeMillis() }.map { it.roomId }.distinct().size
                        val occupancy = if (activeRooms == 0) 0 else occupiedRoomIds * 100 / activeRooms
                        val monthStart = java.util.Calendar.getInstance().apply { set(java.util.Calendar.DAY_OF_MONTH, 1); set(java.util.Calendar.HOUR_OF_DAY, 0) }.timeInMillis
                        findViewById<TextView>(R.id.tv_operational_summary).text =
                            "Phòng đang mở: $activeRooms/${rooms.size}  •  Đang lưu trú: $occupiedRoomIds\nNgười dùng mới tháng này: ${users.count { it.createdAt >= monthStart }}  •  Tỷ lệ lấp đầy hiện tại: $occupancy%"
                        val dateFormat = SimpleDateFormat("dd/MM HH:mm", Locale("vi", "VN"))
                        findViewById<TextView>(R.id.tv_recent_bookings).text = bookings.sortedByDescending { it.createdAt }.take(5)
                            .joinToString("\n") { "#${it.id} • ${dateFormat.format(Date(it.createdAt))} • ${statusLabel(it.status)} • ${NumberFormat.getNumberInstance(Locale("vi", "VN")).format(it.totalPrice.toLong())} đ" }
                            .ifBlank { "Chưa có booking" }
                    }
            }
        }
    }

    private fun statusLabel(status: String) = when (status) {
        "pending" -> "Chờ duyệt"; "confirmed" -> "Đã xác nhận"
        "completed" -> "Hoàn thành"; "cancelled" -> "Đã hủy"; else -> status
    }
    
    private fun setupViews() {
        val tvAdminName = findViewById<TextView>(R.id.tv_admin_name)
        val cardManageRooms = findViewById<MaterialCardView>(R.id.card_manage_rooms)
        val cardViewUsers = findViewById<MaterialCardView>(R.id.card_view_users)
        val cardManageBookings = findViewById<MaterialCardView>(R.id.card_manage_bookings)
        val btnLogout = findViewById<MaterialButton>(R.id.btn_logout)
        val btnSwitchToGuest = findViewById<MaterialButton>(R.id.btn_switch_to_guest)
        
        // Load admin name
        val prefs = getSharedPreferences("AdminSession", MODE_PRIVATE)
        val adminName = prefs.getString("admin_fullname", "Administrator")
        tvAdminName.text = adminName
        
        // Quản lý Phòng
        cardManageRooms.setOnClickListener {
            startActivity(Intent(this, AdminRoomsActivity::class.java))
        }
        
        // Xem Users
        cardViewUsers.setOnClickListener {
            startActivity(Intent(this, AdminUsersActivity::class.java))
        }
        
        // Quản lý Bookings
        cardManageBookings.setOnClickListener {
            startActivity(Intent(this, AdminBookingsActivity::class.java))
        }

        btnSwitchToGuest.setOnClickListener {
            btnSwitchToGuest.isEnabled = false
            lifecycleScope.launch {
                ensureAdminCustomerSession()
                startActivity(Intent(this@AdminDashboardActivity, MainActivity::class.java).apply {
                    putExtra("admin_guest_preview", true)
                    flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                })
                finish()
            }
        }
        
        // Logout
        btnLogout.setOnClickListener {
            clearAdminSession()
            Toast.makeText(this, "Đã đăng xuất", Toast.LENGTH_SHORT).show()
            val intent = Intent(this, LoginActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            startActivity(intent)
            finish()
        }
    }
    
    private fun isAdminLoggedIn(): Boolean {
        val prefs = getSharedPreferences("AdminSession", MODE_PRIVATE)
        return prefs.getBoolean("is_admin_logged_in", false)
    }
    
    private fun clearAdminSession() {
        val prefs = getSharedPreferences("AdminSession", MODE_PRIVATE)
        prefs.edit().clear().apply()
        SessionManager(this).clearSession()
    }

    private suspend fun ensureAdminCustomerSession() {
        val session = SessionManager(this)
        if (session.isLoggedIn()) return

        val repository = (application as HomestayApplication).repository
        val adminUser = repository.getUserByEmail(AdminAuth.EMAIL) ?: run {
            val id = repository.insertUser(
                User(
                    email = AdminAuth.EMAIL,
                    phone = "admin-local",
                    password = "AdminLocalSession@1",
                    fullName = "Administrator"
                )
            )
            repository.getUserById(id) ?: return
        }
        session.saveSession(
            userId = adminUser.id,
            mongoUserId = "local-admin",
            email = adminUser.email,
            name = adminUser.fullName
        )
    }
}

